crates/acvm/src/lib.rs

```diff
@@ -38,178 +39,211 @@ pub trait Backend: SmartContract + ProofSystemCompiler + PartialWitnessGenerator
+                },
+                Directive::Oddrange { a, b, r, bit_size } => match initial_witness.get(a) {
+                    Some(val_a) => {
+                        let int_a = BigUint::from_bytes_be(&val_a.to_bytes());
+                        let pow: BigUint = BigUint::one() << (bit_size - 1);
+                        if int_a >= (&pow << 1) {
+                            return GateResolution::UnsatisfiedConstrain;
                         }
-                        _ => true,
-                    },
+                        let bb = &int_a & &pow;
+                        let int_r = &int_a - &bb;
+                        let int_b = &bb >> (bit_size - 1);
+
+                        initial_witness
+                            .insert(*b, FieldElement::from_be_bytes_reduce(&int_b.to_bytes_be()));
+                        initial_witness
+                            .insert(*r, FieldElement::from_be_bytes_reduce(&int_r.to_bytes_be()));
+                        GateResolution::Resolved
+                    }
+                    _ => GateResolution::Skip,
                 },
+            },
+        }
+    }
+
+    fn solve(
+        &self,
+        initial_witness: &mut BTreeMap<Witness, FieldElement>,
+        mut gates: Vec<Gate>,
+    ) -> GateResolution {
+        let mut gates2: Vec<Gate> = Vec::new();
+        let mut pass_nb = 0;
+        let mut gates_to_resolve = &mut gates;
+        let mut unresolved_gates = &mut gates2;
+        let mut ctx = BinarySolver::new();
+        let mut binary_solve = -1;
+
+        while !gates_to_resolve.is_empty() {
+            unresolved_gates.clear();
+
+            let mut process = |gate| {
+                let mut result = self.solve_gate(initial_witness, gate);
+                if binary_solve >= 0 && result == GateResolution::Skip {
+                    result = ctx.solve(gate, initial_witness);
+                }
+                result
             };
-            if unsolved {
-                unsolved_gates.push(gate);
+
+            if binary_solve != 0 {
+                for gate in gates_to_resolve.iter() {
+                    let result = process(gate);
+                    match result {
+                        GateResolution::Skip => unresolved_gates.push(gate.clone()),
+                        GateResolution::Resolved => (),
+                        _ => return result,
+                    }
+                }
+            } else {
+                //we go backward because binary solver should execute only when the program returns an array
+                //in that case it is a bit more efficient to go backwards, although both ways work.
+                for gate in gates_to_resolve.iter().rev() {
+                    let result = process(gate);
+                    match result {
+                        GateResolution::Skip => unresolved_gates.push(gate.clone()),
+                        GateResolution::Resolved => (),
+                        _ => return result,
```

Comment:
I wanted to do this but it does not work: `if` and `else` have incompatible types

Reply:
I think we have 2 options here:
- We can create a &dyn variable:
```rs
let gates: Box<dyn Iterator<Item = _>> = if binary_solve == Some(0) {
        Box::new(gates_to_resolve.iter())
    } else {
        Box::new(gates_to_resolve.iter().rev())
    };
```
- Or we can factor out the for loop into a helper function. It can't be a closure unfortunately since those cannot be polymorphic. The free function would need to return an Optional value so we could check for that and early return if so.

Unfortunate we have to go a bit out of our way in either case but I do think it is worth it to avoid repeating the block.